packages/contracts-bedrock/scripts/ChainAssertions.sol

```diff
@@ -0,0 +1,187 @@
+            ProtocolVersions: protocolVersions
+        });
+        postDeployAssertions(proxies, cfg, l2OutputOracleStartingTimestamp, vm);
+    }
+
+    /// @notice Asserts the correctness of an L1 deployment
+    function postDeployAssertions(
+        Types.ContractSet memory prox,
+        DeployConfig cfg,
+        uint256 l2OutputOracleStartingTimestamp,
+        Vm vm
+    )
+        internal
+        view
+    {
+        ResourceMetering.ResourceConfig memory rcfg = SystemConfig(prox.SystemConfig).resourceConfig();
+        ResourceMetering.ResourceConfig memory dflt = Constants.DEFAULT_RESOURCE_CONFIG();
+        require(keccak256(abi.encode(rcfg)) == keccak256(abi.encode(dflt)));
+
+        checkSystemConfig(prox, cfg);
+        checkL1CrossDomainMessenger(prox, vm);
+        checkL1StandardBridge(prox, vm);
+        checkL2OutputOracle(prox, cfg, l2OutputOracleStartingTimestamp);
+        checkOptimismMintableERC20Factory(prox);
+        checkL1ERC721Bridge(prox);
+        checkOptimismPortal(prox, cfg);
+        checkProtocolVersions(prox, cfg);
+    }
+
+    /// @notice Asserts that the SystemConfig is setup correctly
+    function checkSystemConfig(Types.ContractSet memory proxies, DeployConfig cfg) internal view {
+        checkSystemConfigV0(proxies.SystemConfig, cfg);
+
+        // Then check the non-legacy fields
+        SystemConfig config = SystemConfig(proxies.SystemConfig);
+        require(config.l1ERC721Bridge() == proxies.L1ERC721Bridge);
+        require(config.l1StandardBridge() == proxies.L1StandardBridge);
+        require(config.l2OutputOracle() == proxies.L2OutputOracle);
+        require(config.optimismPortal() == proxies.OptimismPortal);
+        require(config.l1CrossDomainMessenger() == proxies.L1CrossDomainMessenger);
+
+        // A non zero start block is an override
+        uint256 startBlock = cfg.systemConfigStartBlock();
+        if (startBlock != 0) {
+            require(config.startBlock() == startBlock);
+        } else {
+            require(config.startBlock() == block.number);
+        }
+    }
+
+    /// @notice Asserts that the legacy SystemConfig is setup correctly
+    function checkSystemConfigV0(address systemConfigProxy, DeployConfig cfg) internal view {
+        ISystemConfigV0 config = ISystemConfigV0(systemConfigProxy);
+        require(config.owner() == cfg.finalSystemOwner());
+        require(config.overhead() == cfg.gasPriceOracleOverhead());
+        require(config.scalar() == cfg.gasPriceOracleScalar());
+        require(config.batcherHash() == bytes32(uint256(uint160(cfg.batchSenderAddress()))));
+        require(config.unsafeBlockSigner() == cfg.p2pSequencerAddress());
+
+        ResourceMetering.ResourceConfig memory rconfig = Constants.DEFAULT_RESOURCE_CONFIG();
+        ResourceMetering.ResourceConfig memory resourceConfig = config.resourceConfig();
+        require(resourceConfig.maxResourceLimit == rconfig.maxResourceLimit);
```

Comment:
I am realizing another way to do this would be to abi encode and hash the structs, this ensures future changes don't need a diff here but then we lose some readability. What do you think? I lean towards hashing the struct for safety

Reply:
Only issue I have with a hash comparison is it creates a false positive once the `DeployConfig`. Which I can see happening as MCP deployments deviate from this library.
Now that ChainAssertions are broken up, it could be better to simply have different versions for `SystemConfig` checks? Leaving the caller the responsibility of choosing the appropriate one based on their deployment.